python/cudf/cudf/core/tools/datetimes.py

```diff
@@ -814,23 +842,17 @@ def date_range(
     if isinstance(freq, DateOffset):
         offset = freq
     elif isinstance(freq, str):
-        # Map pandas `offset alias` into cudf DateOffset `CODE`, only
-        # fixed-frequency, non-anchored offset aliases are supported.
-        mo = re.fullmatch(
-            rf'(-)*(\d*)({"|".join(_offset_alias_to_code.keys())})', freq
+        e = ValueError(
+            f"Unrecognized frequency string {freq}. cuDF does"
+            " not yet support month, quarter, year-anchored frequency."
         )
-        if mo is None:
-            raise ValueError(
-                f"Unrecognized or unsupported offset alias {freq}."
-            )
 
-        sign, n, offset_alias = mo.groups()
-        code = _offset_alias_to_code[offset_alias]
-
-        freq = "".join([n, code])
-        offset = DateOffset._from_freqstr(freq)
-        if sign:
-            offset.kwds.update({s: -i for s, i in offset.kwds.items()})
+        if "M" in freq or "Y" in freq.upper():
+            raise e
```

Comment:
This `if` clause is a bit of an awkward special-case for supporting `ms`. If the `DateOffset._from_str` will raise a `ValueError` here, I would just remove the `if: raise` logic that exits early here and let it be handled in the except block.

Reply:
`DateOffset._from_str` does not raise for `M` and `Y`. Which is why it needs to get handled ahead here. It feels like `_from_str` can only handle so many common paths. `M` means months for a DateOffset but "Months End" in `date_range` context. I meant to support constructing months with string like `1M` in https://github.com/rapidsai/cudf/pull/9710 but reject it here. The two functions handles `M` in conflicting meanings so this is as far as I can do to consolidate the logic.

Reply:
Can you add a comment that clarifies which paths you're treating as special cases here? Something that specifies how DateOffset and date_range differ in functionality and that you're intending to catch month/year with this statement.